src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/flow_units/ClusterResourceFlowUnit.java

```diff
@@ -0,0 +1,99 @@
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.ResourceEnum;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.contexts.ResourceContext;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.HotClusterSummary;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.HotNodeSummary;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.HotResourceSummary;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.bucket.BucketCalculator;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.bucket.UsageBucket;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.RcaConf;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.store.rca.cluster.NodeKey;
+
+/**
+ * Base class for cluster level {@link ResourceFlowUnit}s which want to bucketize resources into
+ * various {@link UsageBucket}s
+ *
+ * <p>If you'd like to subclass this class, simply implement initBucketCalculator() to return an
+ * appropriate {@link BucketCalculator} to use for bucketization. {@link CpuClusterResourceFlowUnit}
+ * is a good example of this.
+ */
+public class ClusterResourceFlowUnit extends ResourceFlowUnit<HotClusterSummary> {
+  // This config is useful for reading bucket ranges from the rca.conf file
+  protected RcaConf rcaConf;
+  // This calculator is used to bucketize our resources per node
+  protected BucketCalculator bucketCalculator;
+
+  public ClusterResourceFlowUnit(long timeStamp) {
+    super(timeStamp);
+  }
+
+  public ClusterResourceFlowUnit(long timeStamp, ResourceContext context,
+      HotClusterSummary summary, RcaConf rcaConf, boolean persistSummary) {
+    super(timeStamp, context, summary, persistSummary);
+    this.rcaConf = rcaConf;
+    this.bucketCalculator = initBucketCalculator();
+    computeUsageBuckets();
+  }
+
+  /**
+   * Returns the {@link BucketCalculator} to use for bucketization of resources
+   *
+   * @return the {@link BucketCalculator} to use for bucketization of resources
+   */
+  protected BucketCalculator initBucketCalculator() {
+    return null;
+  }
+
+  /**
+   * Computes which usage bucket each resource in this class's summary should belong to.
+   *
+   * <p>You probably don't need to override this. Just implement initBucketCalculator()
+   */
+  protected void computeUsageBuckets() {
+    if (getSummary() == null) {
+      return;
+    }
+    for (HotNodeSummary hotNodeSummary : getSummary().getHotNodeSummaryList()) {
+      for (HotResourceSummary hotResourceSummary : hotNodeSummary.getHotResourceSummaryList()) {
+        NodeKey nodeKey = new NodeKey(hotNodeSummary.getNodeID(), hotNodeSummary.getHostAddress());
+        if (bucketCalculator != null) {
```

Comment:
If we dont do anything when `bucketCalculator == null` then we can move this check to the beginning of the fn.